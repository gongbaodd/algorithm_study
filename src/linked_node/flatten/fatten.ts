import { LNode } from "./lib";

function merge(head1: LNode | null, head2: LNode | null) {
  if (!head1) {
    return head2;
  }

  if (!head2) {
    return head1;
  }

  let result: LNode;

  if (head1.data < head2.data) {
    result = head1;
    result.down = merge(head1.down, head2);
  } else {
    result = head2;
    result.down = merge(head1, head2.down);
  }
  return result;
}

export function flatten(_head: LNode): LNode | null {
  let head: LNode | null = _head;
  if (head.right === null) {
    return head;
  }

  if (head.down === null) {
    return head;
  }

  head.right = flatten(head.right);
  head = merge(head, head.right);

  return head;
}
