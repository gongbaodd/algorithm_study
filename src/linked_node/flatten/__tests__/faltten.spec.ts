import { init } from "../lib";
import { flatten } from "../fatten";

describe("flatten", () => {
  it("should return [3,6,8,11,15,21,22,30,31,39,40,45,50]", () => {
    const head = flatten(init());
    expect(head).toMatchSnapshot();
  });
});
