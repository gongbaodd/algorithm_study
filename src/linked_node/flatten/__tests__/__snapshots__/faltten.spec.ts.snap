// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flatten should return [3,6,8,11,15,21,22,30,31,39,40,45,50] 1`] = `
LNode {
  "data": 3,
  "down": LNode {
    "data": 6,
    "down": LNode {
      "data": 8,
      "down": LNode {
        "data": 11,
        "down": LNode {
          "data": 15,
          "down": LNode {
            "data": 21,
            "down": LNode {
              "data": 22,
              "down": LNode {
                "data": 30,
                "down": LNode {
                  "data": 31,
                  "down": LNode {
                    "data": 39,
                    "down": LNode {
                      "data": 40,
                      "down": LNode {
                        "data": 50,
                        "down": LNode {
                          "data": 55,
                          "down": null,
                          "right": null,
                        },
                        "right": null,
                      },
                      "right": null,
                    },
                    "right": null,
                  },
                  "right": null,
                },
                "right": null,
              },
              "right": null,
            },
            "right": null,
          },
          "right": LNode {
            "data": 30,
            "down": LNode {
              "data": 31,
              "down": LNode {
                "data": 39,
                "down": LNode {
                  "data": 40,
                  "down": LNode {
                    "data": 50,
                    "down": LNode {
                      "data": 55,
                      "down": null,
                      "right": null,
                    },
                    "right": null,
                  },
                  "right": null,
                },
                "right": null,
              },
              "right": null,
            },
            "right": null,
          },
        },
        "right": LNode {
          "data": 15,
          "down": LNode {
            "data": 21,
            "down": LNode {
              "data": 22,
              "down": LNode {
                "data": 30,
                "down": LNode {
                  "data": 31,
                  "down": LNode {
                    "data": 39,
                    "down": LNode {
                      "data": 40,
                      "down": LNode {
                        "data": 50,
                        "down": LNode {
                          "data": 55,
                          "down": null,
                          "right": null,
                        },
                        "right": null,
                      },
                      "right": null,
                    },
                    "right": null,
                  },
                  "right": null,
                },
                "right": null,
              },
              "right": null,
            },
            "right": null,
          },
          "right": LNode {
            "data": 30,
            "down": LNode {
              "data": 31,
              "down": LNode {
                "data": 39,
                "down": LNode {
                  "data": 40,
                  "down": LNode {
                    "data": 50,
                    "down": LNode {
                      "data": 55,
                      "down": null,
                      "right": null,
                    },
                    "right": null,
                  },
                  "right": null,
                },
                "right": null,
              },
              "right": null,
            },
            "right": null,
          },
        },
      },
      "right": null,
    },
    "right": null,
  },
  "right": LNode {
    "data": 11,
    "down": LNode {
      "data": 15,
      "down": LNode {
        "data": 21,
        "down": LNode {
          "data": 22,
          "down": LNode {
            "data": 30,
            "down": LNode {
              "data": 31,
              "down": LNode {
                "data": 39,
                "down": LNode {
                  "data": 40,
                  "down": LNode {
                    "data": 50,
                    "down": LNode {
                      "data": 55,
                      "down": null,
                      "right": null,
                    },
                    "right": null,
                  },
                  "right": null,
                },
                "right": null,
              },
              "right": null,
            },
            "right": null,
          },
          "right": null,
        },
        "right": null,
      },
      "right": LNode {
        "data": 30,
        "down": LNode {
          "data": 31,
          "down": LNode {
            "data": 39,
            "down": LNode {
              "data": 40,
              "down": LNode {
                "data": 50,
                "down": LNode {
                  "data": 55,
                  "down": null,
                  "right": null,
                },
                "right": null,
              },
              "right": null,
            },
            "right": null,
          },
          "right": null,
        },
        "right": null,
      },
    },
    "right": LNode {
      "data": 15,
      "down": LNode {
        "data": 21,
        "down": LNode {
          "data": 22,
          "down": LNode {
            "data": 30,
            "down": LNode {
              "data": 31,
              "down": LNode {
                "data": 39,
                "down": LNode {
                  "data": 40,
                  "down": LNode {
                    "data": 50,
                    "down": LNode {
                      "data": 55,
                      "down": null,
                      "right": null,
                    },
                    "right": null,
                  },
                  "right": null,
                },
                "right": null,
              },
              "right": null,
            },
            "right": null,
          },
          "right": null,
        },
        "right": null,
      },
      "right": LNode {
        "data": 30,
        "down": LNode {
          "data": 31,
          "down": LNode {
            "data": 39,
            "down": LNode {
              "data": 40,
              "down": LNode {
                "data": 50,
                "down": LNode {
                  "data": 55,
                  "down": null,
                  "right": null,
                },
                "right": null,
              },
              "right": null,
            },
            "right": null,
          },
          "right": null,
        },
        "right": null,
      },
    },
  },
}
`;
